python/ray/ml/train/data_parallel_trainer.py

```diff
@@ -181,4 +189,141 @@ def __init__(
+        checkpoint_manager.on_init(preprocessor=self.preprocessor)
+
+        # Start the remote actors.
+        ray.get(backend_executor_actor.start.remote(initialization_hook=None))
+
+        if self.resume_from_checkpoint:
+            resume_checkpoint_dict = self.resume_from_checkpoint.to_dict()
+        else:
+            resume_checkpoint_dict = None
+
+        # Tell Ray Train to only shard the train dataset and not the other datasets.
+        # This is purely an implementation detail and users do not need to know about
+        # this.
+        # TODO(amog): Refactor this to remove hack and make this more modular.
+        #  TrainingIterator should accept a generic custom_ingest_func that contains
+        #  the logic for how to split the Datasets.
+        updated_dataset_dict = {}
+        for key, value in self.datasets.items():
+            if key == TRAIN_DATASET_KEY:
+                updated_dataset_dict[key] = value
+            else:
+                # Ray Train will strip out the added string before exposing to users.
+                updated_dataset_dict[key + "_NO-SHARD"] = value
+
+        # TODO(amog): Have TrainingIterator also accept a checkpoint ObjectRef instead
+        #  of just a Dict.
+        training_iterator = TrainingIterator(
+            backend_executor_actor=backend_executor_actor,
+            backend_config=self.backend_config,
+            train_func=train_loop_per_worker,
+            dataset=updated_dataset_dict if len(updated_dataset_dict) > 0 else None,
+            checkpoint_manager=checkpoint_manager,
+            checkpoint=resume_checkpoint_dict,
+            checkpoint_strategy=None,
+        )
+
+        for results in training_iterator:
+            # TODO(ml-team): add ability to report results from multiple workers.
+            first_worker_results = results[0]
+
+            tune.report(**first_worker_results)
+
+        # Shutdown workers.
+        ray.get(backend_executor_actor.shutdown.remote())
+
+
+# TODO(team-ml): Refactor checkpoint management along with Tune.
+class _DataParallelCheckpointManager(TuneCheckpointManager):
+    def on_init(self, preprocessor: Preprocessor):
+        self.preprocessor = preprocessor
+        super(_DataParallelCheckpointManager, self).on_init()
+
+    def write_checkpoint(self, checkpoint: Dict):
+        self.add_tune_checkpoint_id(checkpoint)
+
+        # Add the preprocessor to the checkpoint.
+        checkpoint[PREPROCESSOR_KEY] = self.preprocessor
+
+        checkpoint_obj = Checkpoint.from_dict(checkpoint)
+        # If inside a Tune Trainable, then checkpoint with Tune.
+        with tune.checkpoint_dir(step=self._latest_checkpoint_id) as checkpoint_dir:
+            checkpoint_obj.to_directory(path=checkpoint_dir)
+
+    @property
+    def latest_checkpoint_dir(self) -> Optional[Path]:
+        raise RuntimeError(
+            "`latest_checkpoint_dir` should not be called when running " "with Tune."
+        )
```

Comment:
Add the same for the other properties? Also I was thinking we could just raise  `NotImplementedError`.

Reply:
This is a more general problem with `TuneCheckpointManager`, in general. I would say let's leave this for a future PR when we can clean all this up!